File: java/sieve/src/main/java/Sieve.java
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Sieve {
    private final int upTo;

    public Sieve(int upTo) {
        this.upTo = upTo;
    }

    public List<Integer> getPrimes() {
        List<Integer> possiblePrimes = IntStream.range(2, this.upTo + 1).boxed().collect(Collectors.toList());
        List<Integer> primes = new LinkedList<>();

        while (possiblePrimes.size() > 0) {
            int prime = possiblePrimes.get(0);
            primes.add(prime);

            possiblePrimes = excludeMultiples(prime, possiblePrimes);
        }

        return primes;
    }

    private static List<Integer> excludeMultiples(int prime, List<Integer> possiblePrimes) {
        return possiblePrimes
                .stream()
                .filter(x -> x % prime != 0)
                .collect(Collectors.toList());
    }
}

File: ecmascript/sieve/sieve.js
function* range(begin, end) {
    for (let i = begin; i <= end; i++) {
        yield i;
    }
}

export default class Sieve {
    constructor(max) {
        this.max = max;
    }

    get primes() {
        let possiblePrimes = [...range(2, this.max)];
        const primes = [];

        while (possiblePrimes.length > 0) {
            const prime = possiblePrimes[0];
            primes.push(prime);

            possiblePrimes = this.excludePrimeMultiples(possiblePrimes, prime);
        }

        return primes;
    }

    excludePrimeMultiples(possiblePrimes, prime) {
        return possiblePrimes.filter(n => n % prime !== 0);
    }
}